packages/sync-service/lib/electric/shape_cache/pure_file_storage.ex

```
@@ -892,6 +910,9 @@ defmodule Electric.ShapeCache.PureFileStorage do
     {latest_name, {compaction_boundary, compacted_name}, {cached_min, chunks}} =
       boundary_info || get_read_source_info(opts)
 
+    # Any virtual offsets are handled elsewhere - normalize them to the last before real offsets for main log
+    offset = if is_virtual_offset(offset), do: LogOffset.last_before_real_offsets(), else: offset
```

this confuses me slightly - if virtual offsets are handled elsewhere, and this is a private function, how do we ever end up handling a virtual offset here, and how do we know its safe to cast it into a last before real offsets? Should we not just add a guard to the function and let it fail if we don't expect it to ever get here? (or fix whatever code path leads here with a virtual offset?)

---

It was normalized to 0,inf in one call path but not the other. I prefer it normalized here in one place